lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -3771,77 +3640,989 @@ inline llvm::raw_ostream &operator<<(llvm::raw_ostream &os,
+
+    // TODO: Generalize for indirect results, multiple results, etc
+    auto origResult = ai->getResult(indices.source);
+
+    // Extract all direct results from the differential.
+    SmallVector<SILValue, 8> differentialDirResults;
+    extractAllElements(differentialCall, diffBuilder, differentialDirResults);
+    // Get all differential results in type-defined order.
+    SmallVector<SILValue, 8> differentialAllResults;
+    collectAllActualResultsInTypeOrder(
+        differentialCall, differentialDirResults,
+        differentialCall->getIndirectSILResults(), differentialAllResults);
+    auto differentialResult = differentialAllResults[indices.source];
+
+    // Add tangent for original result.
+    assert(indices.source == 0 && "Expected result index to be first.");
+    setTangentValue(bb, origResult,
+                    makeConcreteTangentValue(differentialResult));
+
+    // TODO: handle indirect
+    // Add the tangent value of the result of the apply instruction.
+    //    auto tanVal = makeConcreteAdjointValue(
+    //        ValueWithCleanup(originalDirectResult,
+    //        makeCleanup(originalDirectResult, emitCleanup, {})));
+    //    for (auto dirRes : ai->getResults())
+    //      addTangentValue(bb, originalDirectResult, tanVal);
+    //    for (auto indRes : ai->getIndirectSILResults())
+    //      ;
+  }
+
+public:
+  explicit JVPEmitter(ADContext &context, SILFunction *original,
+                      SILDifferentiableAttr *attr, SILFunction *jvp,
+                      DifferentiationInvoker invoker)
+      : TypeSubstCloner(*jvp, *original, getSubstitutionMap(original, jvp)),
+        context(context), original(original), attr(attr), jvp(jvp),
+        differentialInfo(context, AutoDiffAssociatedFunctionKind::JVP, original,
+                         jvp, attr->getIndices()), invoker(invoker),
+        activityInfo(
+            getActivityInfo(context, original, attr->getIndices(), jvp)),
+        differentialAndBuilder(initializeDifferentialAndBuilder(
+            context, original, attr, &differentialInfo)),
+        diffLocalAllocBuilder(getDifferential()) {
+    // Get JVP generic signature.
+    CanGenericSignature jvpGenSig = nullptr;
+    if (auto *jvpGenEnv = jvp->getGenericEnvironment())
+      jvpGenSig = jvpGenEnv->getGenericSignature()->getCanonicalSignature();
+    // Create empty differential function.
+    context.getGeneratedFunctions().push_back(&getDifferential());
+  }
+
+  static SILFunction *createEmptyDifferential(
+      ADContext &context, SILFunction *original, SILDifferentiableAttr *attr,
+      LinearMapInfo *linearMapInfo) {
+    auto &module = context.getModule();
+    auto origTy = original->getLoweredFunctionType();
+    auto lookupConformance = LookUpConformanceInModule(module.getSwiftModule());
+
+    // RAII that pushes the original function's generic signature to
+    // `module.Types` so that the calls `module.Types.getTypeLowering()` below
```

Comment:
```suggestion
    // `module.Types` so that calls to `module.Types.getTypeLowering()` below
```
Rename everywhere please!